### PR TITLE
cleanFeature/nino/clean

### PR DESCRIPTION
**Pull request by jujoru**
OBSERVATIONS
I have used the base of the Juan Antonio branch project to optimize development times.
I have changed scaffolding, names, structures, creating new interfaces and class, etc ...

SUMMARY
Migration to Clean Architecture

CHANGES MADE
Layers of architecture by modules
Use LocationService to provider repository
Mapper
Creating interface IUseCase for implements in UsesCase

### CODE REVIEW

File: usecases/src/main/java/com/architectcoders/usecases/GetCharacters.kt [hunk 1]
```diff
@@ -3,6 +3,14 @@ package com.architectcoders.usecases
...
-class GetCharacters(private val characterRepository: CharacterRepository) {
-    suspend fun invoke(offset: Int): Characters = characterRepository.getCharactersRemote(offset)
+class GetCharacters(private val characterRepository: CharacterRepository): IUseCase<Characters,Int>{
+    override suspend fun  invoke(params: Int): Characters  =
+           characterRepository.getCharactersRemote(params)
+
+
+
+
+    // suspend fun invoke(offset: Int): Characters = characterRepository.getCharactersRemote(offset)
```
Comment (juancoob) [inline usecases/src/main/java/com/architectcoders/usecases/GetCharacters.kt]: @jujoru could you remove the commented code?

File: usecases/src/main/java/com/architectcoders/usecases/IUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.architectcoders.usecases
+
+import com.architectcoders.data.repository.CharacterRepository
+import com.architectcoders.domain.characters.Characters
+
+interface IUseCase <T,K>{
+ suspend fun invoke(param: K): T?
```
Comment (juancoob) [inline usecases/src/main/java/com/architectcoders/usecases/IUseCase.kt]: @jujoru that's awesome! shouldn't be better if we respect the order of parameters? I mean, T as a function parameter and K as a returned object

**Review by juancoob — APPROVED** [review]
Thanks @jujoru !! Could you check my comments?

File: usecases/src/main/java/com/architectcoders/usecases/GetCharacters.kt (hunk 1 above)
Comment (jujoru) [reply usecases/src/main/java/com/architectcoders/usecases/GetCharacters.kt]: Yes ¡¡¡ thanks.

File: usecases/src/main/java/com/architectcoders/usecases/IUseCase.kt (hunk 2 above)
Comment (jujoru) [reply usecases/src/main/java/com/architectcoders/usecases/IUseCase.kt]: Perfect, I change this.

**Review by joseantoniocampillo — APPROVED** [review]
That's good

**Pull request merged**